src/adventure/adventuretracker.cpp

```diff
@@ -0,0 +1,179 @@
+#include "parser/parserutils.h"
+
+#include <QDebug>
+
+AdventureTracker::AdventureTracker(GameObserver &observer, QObject *const parent)
+    : QObject{parent}
+    , m_observer{observer}
+    , m_session{nullptr}
+{
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserText,
+            this,
+            &AdventureTracker::slot_onUserText);
+
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserGmcp,
+            this,
+            &AdventureTracker::slot_onUserGmcp);
+}
+
+void AdventureTracker::slot_onUserText(const QByteArray &ba)
+{
+    QString line = QString::fromLatin1(ba).trimmed();
+    ParserUtils::removeAnsiMarksInPlace(line);
+
+    // Try to order these by frequency to minimize unnecessary parsing
+
+    if (m_killParser.parse(line)) {
+        auto killName = m_killParser.getLastSuccessVal();
+        double xpGained = checkpointXP();
+        emit sig_killedMob(killName, xpGained);
+        return;
+    }
+
+    if (m_gainedLevelParser.parse(line)) {
+        emit sig_gainedLevel();
+        return;
+    }
+
+    if (m_achievementParser.parse(line)) {
+        auto achievement = m_achievementParser.getLastSuccessVal();
+        auto xpGained = checkpointXP();
+        emit sig_achievedSomething(achievement, xpGained);
+        return;
+    }
+
+    if (m_accomplishedTaskParser.parse(line)) {
+        auto xpGained = checkpointXP();
+        emit sig_accomplishedTask(xpGained);
+        return;
+    }
+
+    if (m_diedParser.parse(line)) {
+        auto xpLost = checkpointXP();
+        emit sig_died(xpLost);
+        return;
+    }
+
+    if (m_hintParser.parse(line)) {
+        auto hint = m_hintParser.getLastSuccessVal();
+        emit sig_receivedHint(hint);
+        return;
+    }
+}
+
+void AdventureTracker::slot_onUserGmcp(const GmcpMessage &msg)
+{
+    // https://mume.org/help/generic_mud_communication_protocol
+
+    if (msg.isCharName()) {
+        parseIfUpdatedChar(msg);
+    }
+
+    if (msg.isCharStatusVars()) {
+        parseIfUpdatedChar(msg);
+    }
+
+    if (msg.isCharVitals()) {
+        parseIfUpdatedXP(msg);
+    }
+
+    if (msg.isCommChannelText()) {
+        parseIfReceivedComm(msg);
+    }
+
+    if (msg.isCoreGoodbye()) {
+        parseIfGoodbye(msg);
+    }
+}
+
+void AdventureTracker::parseIfGoodbye([[maybe_unused]] GmcpMessage msg)
+{
+    // REVISIT If you try to call msg.getJson()->toQString().toUtf8() on a
+    // CoreGoodbye message, the GmcpMessage code crashes, trying to malloc a
+    // huge 0xffffffff chunk of memory. Needs investigation.
+
+    if (m_session != nullptr) {
+        qDebug().noquote() << QString("Adventure: ending session for %1").arg(m_session->name());
+        m_session->endSession();
+        emit sig_endedSession(*m_session);
+        m_session.reset();
+    }
+}
+
+void AdventureTracker::parseIfReceivedComm(GmcpMessage msg)
+{
+    auto s = msg.getJson()->toQString().toUtf8();
+    QJsonObject obj = QJsonDocument::fromJson(s).object();
+
```

Comment:
I guess we're at the point where we might be duplicating work here too. Maybe we need a new class for GMCP messages that are parsed. Perhaps an optional member?

Reply:
I struggled with this. IIUC, some of the GMCP messages are encoded objects, some are encoded arrays of objects, and some are just strings ( I think, like `Comm.Channel.Enable`) and some don't have message bodies. So basically we can't just do a blanket .object() across the board. So either we need to:
- Parse the QJsonDocument in a common class, but still let the callers decide what to extract from the document
- Materialize logic for the full GMCP grammar into the class, so that it knows how to parse different message types

Also, there are dragons here: as I commented elsewhere, if you call msg.getJson()->toQString() on a Core.Goodbye message, it crashes (probably because there is no message body and it's seeking endlessly for a terminating character, or something), so I backed off. We probably need a more robust testing strategy for GMCP in general (or at least a strategy for saving a raw message off the wire and then testing it locally).

I didn't feel comfortable diving into the Gmcp code so I felt this little duplication was a reasonable compromise, but let me know.